### PR TITLE
errorhandling 4583

### PR DESCRIPTION
**Pull request by jdevcs**
## Description

For: https://github.com/ChainSafe/web3.js/issues/4583

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

### CODE REVIEW

File: packages/web3-eth/src/errors.ts [hunk 1]
```diff
@@ -29,15 +29,27 @@ export class InvalidTransactionWithSender extends Web3Error {
 	public code = ERR_TX_INVALID_SENDER;
 
 	public constructor(value: unknown) {
-		super(JSON.stringify(value), 'invalid transaction with sender');
+		super(
+			JSON.stringify(
+				value,
+				(_, v) => (typeof v === 'bigint' ? v.toString() : v) as unknown,
+			).replace(/['"]+/g, ''),
```
Comment (nazarhussain) [inline packages/web3-eth/src/errors.ts]: I would suggest to move this logic of parsing data to string into `web3-utils` package and then use here.

**Review by nazarhussain — CHANGES_REQUESTED** [review]

**Review by luu-alex — APPROVED** [review]

**Review by Spacesai1or — DISMISSED** [review]
This `jsonStringify` shouldn't be applied to specific errors, but rather the [Web3Error](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-common/src/errors.ts#L57) class which our errors extend. In this way, the developer doesn't have to think about it every time they introduce a new error (possibly forgetting to do this, causing a runtime error)

I also wouldn't name this functionality `jsonStringify`, since `JSON.stringify` exists and this isn't just handling JSON. Perhaps we can just make a `toString` method that can handle BigInts, objects, Buffers, etc

**Review by nazarhussain — APPROVED** [review]

**Comment by nazarhussain** [comment]
Apart from @spacesailor24  comment LGTM.

**Comment by avkos** [comment]
> This `jsonStringify` shouldn't be applied to specific errors, but rather the [Web3Error](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-common/src/errors.ts#L57) class which our errors extend. In this way, the developer doesn't have to think about it every time they introduce a new error (possibly forgetting to do this, causing a runtime error)
> 
> I also wouldn't name this functionality `jsonStringify`, since `JSON.stringify` exists and this isn't just handling JSON. Perhaps we can just make a `toString` method that can handle BigInts, objects, Buffers, etc

yes. I also think we need to implement it inside the Web3Error class.

**Review by avkos — CHANGES_REQUESTED** [review]
yes. I also think we need to implement it inside the Web3Error class.

**Comment by jdevcs** [comment]
> This `jsonStringify` shouldn't be applied to specific errors, but rather the [Web3Error](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-common/src/errors.ts#L57) class which our errors extend. In this way, the developer doesn't have to think about it every time they introduce a new error (possibly forgetting to do this, causing a runtime error)
> 
> I also wouldn't name this functionality `jsonStringify`, since `JSON.stringify` exists and this isn't just handling JSON. Perhaps we can just make a `toString` method that can handle BigInts, objects, Buffers, etc

[Web3Error](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-common/src/errors.ts#L57)  is already getting [string](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-common/src/errors.ts#L61) instead of `unknown` type and all of its sub classes are passing string, so no need of parsing that.

More good candidate of parsing `unknown` is [here](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-utils/src/errors.ts#L6), 

 I think this another [web3error](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-utils/src/errors.ts#L2) is created in utils package for avoiding cyclic dependency bw utils and common, but @spacesailor24 Why you used it in web3-eth package instead of web3-error from [common](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-common/src/errors.ts#L57), I think it should be changed in Eth.

**Review by avkos — APPROVED** [review]

**Comment by jdevcs** [comment]
@spacesailor24

**Comment by Spacesai1or** [comment]
> > This `jsonStringify` shouldn't be applied to specific errors, but rather the [Web3Error](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-common/src/errors.ts#L57) class which our errors extend. In this way, the developer doesn't have to think about it every time they introduce a new error (possibly forgetting to do this, causing a runtime error)
> > I also wouldn't name this functionality `jsonStringify`, since `JSON.stringify` exists and this isn't just handling JSON. Perhaps we can just make a `toString` method that can handle BigInts, objects, Buffers, etc
> 
> [Web3Error](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-common/src/errors.ts#L57) is already getting [string](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-common/src/errors.ts#L61) instead of `unknown` type and all of its sub classes are passing string, so no need of parsing that.
> 
> More good candidate of parsing `unknown` is [here](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-utils/src/errors.ts#L6),
> 
> I think this another [web3error](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-utils/src/errors.ts#L2) is created in utils package for avoiding cyclic dependency bw utils and common, but @spacesailor24 Why you used it in web3-eth package instead of web3-error from [common](https://github.com/ChainSafe/web3.js/blob/f71d2fe37cf2848d1368f902b376d5fa6cc96714/packages/web3-common/src/errors.ts#L57), I think it should be changed in Eth.

@jdevcs I didn't realize we had two different implementations of `Web3Error`, dependency cycle aside, this does not seem like the best approach and we should investigate only having one class for errors

I think you're right, `Web3Eth` should be using `Web3Error` from `web3-common` and just pass a formatted string. To do this though, we still require formatting potential `unknown` type values into a `string` to be used for the error message

**Review by nikoulai — APPROVED** [review]

**Comment by jdevcs** [comment]
https://github.com/ChainSafe/web3.js/issues/4889

**Pull request merged**